crates/red_knot_python_semantic/src/types/call/bind.rs

```diff
@@ -1,25 +1,50 @@
-use super::{Argument, CallArguments, InferContext, Signature, Type};
+use super::{
+    Argument, CallArguments, CallError, CallOutcome, CallableSignature, InferContext, Signature,
+    Type,
+};
 use crate::db::Db;
 use crate::types::diagnostic::{
-    INVALID_ARGUMENT_TYPE, MISSING_ARGUMENT, PARAMETER_ALREADY_ASSIGNED,
+    INVALID_ARGUMENT_TYPE, MISSING_ARGUMENT, NO_MATCHING_OVERLOAD, PARAMETER_ALREADY_ASSIGNED,
     TOO_MANY_POSITIONAL_ARGUMENTS, UNKNOWN_ARGUMENT,
 };
 use crate::types::signatures::Parameter;
-use crate::types::{todo_type, CallableType, UnionType};
+use crate::types::{CallableType, UnionType};
 use ruff_db::diagnostic::{OldSecondaryDiagnosticMessage, Span};
 use ruff_python_ast as ast;
 use ruff_text_size::Ranged;
 
-/// Bind a [`CallArguments`] against a callable [`Signature`].
+/// Bind a [`CallArguments`] against a [`CallableSignature`].
 ///
 /// The returned [`CallBinding`] provides the return type of the call, the bound types for all
 /// parameters, and any errors resulting from binding the call.
 pub(crate) fn bind_call<'db>(
     db: &'db dyn Db,
     arguments: &CallArguments<'_, 'db>,
-    signature: &Signature<'db>,
+    overloads: &CallableSignature<'db>,
     callable_ty: Type<'db>,
 ) -> CallBinding<'db> {
+    // TODO: This checks every overload. In the proposed more detailed call checking spec [1],
+    // arguments are checked for arity first, and are only checked for type assignability against
+    // the matching overloads. Make sure to implement that as part of separating call binding into
+    // two phases.
```

Comment:
I specifically called out the arity thing (that's step 1 and 2 in the proposed spec), but there are also a number of other wrinkles, like union-argument expansion (step 3), preference for variadic signature matches to variadic arguments (step 4), and special handling for non-fully-static argument types matching multiple arguments (step 5). Not at all urgent, but at some point would be good to broaden this TODO comment even more to cover these things, as long as we haven't yet implemented them.

Reply:
Right, here I wanted to focus on the bit about only type-checking the overloads that arity-match